File: src/test/java/binary/BinaryTest.java
package binary;

import org.junit.Test;
import utils.Binary;

import static extensions.math.BinaryKt.INT_BITS;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class BinaryTest {

    @Test
    public void hasKthBitSet() {
        for (int k = 0; k < INT_BITS; k++) {
            if (k <= 3) {
                assertTrue(Binary.hasKthBitSet(15, k));
            } else {
                assertFalse(Binary.hasKthBitSet(15, k));
            }
        }

        for (int k = 0; k < INT_BITS; k++) {
            assertFalse(Binary.hasKthBitSet(0, k));
        }
        for (int k = 0; k < INT_BITS; k++) {
            assertTrue(Binary.hasKthBitSet(-1, k));
        }
    }

    @Test
    public void positionOfRightmost1Bit() {
        assertEquals(0, Binary.positionOfRightmost1Bit(15));
        assertEquals(1, Binary.positionOfRightmost1Bit(10));
        assertEquals(3, Binary.positionOfRightmost1Bit(8));
        assertEquals(-1, Binary.positionOfRightmost1Bit(0));
        assertEquals(0, Binary.positionOfRightmost1Bit(-5));
        assertEquals(1, Binary.positionOfRightmost1Bit(234));
    }

    @Test
    public void positionOfRightmost0Bit() {
        assertEquals(4, Binary.positionOfRightmost0Bit(15));
        assertEquals(0, Binary.positionOfRightmost0Bit(10));
        assertEquals(0, Binary.positionOfRightmost0Bit(8));
        assertEquals(0, Binary.positionOfRightmost0Bit(0));
        assertEquals(-1, Binary.positionOfRightmost0Bit(-1));
        assertEquals(2, Binary.positionOfRightmost0Bit(-5));
    }
}

File: src/test/kotlin/extensions/math/BinaryTest.kt
package extensions.math

import junit.framework.Assert.assertEquals
import org.junit.Assert
import org.junit.Test

class BinaryTest {

    @Test
    fun isEven() {
        setOf(-10, -4, 0, 2, 4, 6, 20, 100, 128).forEach { Assert.assertTrue(it.isEven()) }
        setOf(-9, -3, -1, 1, 3, 7, 21, 101, 127).forEach { Assert.assertFalse(it.isEven()) }
    }

    @Test
    fun isOdd() {
        setOf(-9, -3, -1, 1, 3, 7, 21, 101, 127).forEach { Assert.assertTrue(it.isOdd()) }
        setOf(-10, -4, 0, 2, 4, 6, 20, 100, 128).forEach { Assert.assertFalse(it.isOdd()) }
    }

    @Test
    fun toBinaryString() {
        assertEquals("00000000000000000000000000000000", 0.toBinaryString())
        assertEquals("0000000000000000", 0.toBinaryString(16))
        assertEquals("00000000", 0.toBinaryString(8))

        assertEquals("00000000000000000000000000000001", 1.toBinaryString())
        assertEquals("00000000000000000000000000000010", 2.toBinaryString())
        assertEquals("00000000000000000000000000001010", 10.toBinaryString())

        assertEquals("00000000000000000000000001100001", 97.toBinaryString())

        assertEquals("00000000000000000000000001000011", 67.toBinaryString())
        assertEquals("11111111111111111111111110111101", (-67).toBinaryString())

        assertEquals("11111111111111111111111111111111", (-1).toBinaryString())
        assertEquals("11111111111111111111111111111110", (-2).toBinaryString())

        assertEquals("01111111111111111111111111111110", (Int.MAX_VALUE - 1).toBinaryString())
        assertEquals("01111111111111111111111111111111", Int.MAX_VALUE.toBinaryString())        // 2147483647

        assertEquals("10000000000000000000000000000001", (Int.MIN_VALUE + 1).toBinaryString())
        assertEquals("10000000000000000000000000000000", Int.MIN_VALUE.toBinaryString())        // -2147483648

        // Overflow
        assertEquals(Int.MIN_VALUE.toBinaryString(), (Int.MAX_VALUE + 1).toBinaryString())
        assertEquals(Int.MAX_VALUE.toBinaryString(), (Int.MIN_VALUE - 1).toBinaryString())
    }
}